File: building_classes/Building.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MiddleAges
{
    class Building : Price
    {
        //#################################################
        // CONSTRUCTOR
        //#################################################
        public Building() { }

        public Building(int _gold, int _wood, int _stone, double _dropRate = 1.0, int _level = 1)
        {
            PriceUpgradeGold = _gold;
            PriceUpgradeWood = _wood;
            PriceUpgradeStone = _stone;
            BuildingDropRate = _dropRate;
            BuildingLevel = _level;
        }

        //#################################################
        // MEMBER FUNCTIONS
        //#################################################
        public void IncreaseDropRate(double _value)
        {
            _buildingDropRate *= _value;
        }

        public void Upgrade(int _gold, int _wood, int _stone)
        {
            _buildingLevel += 1;
            switch (_buildingLevel)
            {
                case 1:
                    IncreaseUpgradePrice(_gold, _wood, _stone, 0, 0);
                    IncreaseDropRate(1);
                    break;
                case 2:
                    IncreaseUpgradePrice(1.1 * _gold, 1.1 * _wood, 1.1 * _stone, 0, 0);
                    IncreaseDropRate(1.1);
                    break;
                case 3:
                    IncreaseUpgradePrice(1.2 * _gold, 1.2 * _wood, 1.2 * _stone, 0, 0);
                    IncreaseDropRate(1.2);
                    break;
                case 4:
                    IncreaseUpgradePrice(1.5 * _gold, 1.5 * _wood, 1.5 * _stone, 0, 0);
                    IncreaseDropRate(1.3);
                    break;
                case 5:
                    IncreaseUpgradePrice(1.8 * _gold, 1.8 * _wood, 1.8 * _stone, 0, 0);
                    IncreaseDropRate(1.4);
                    break;
                default:
                    IncreaseUpgradePrice(0, 0, 0, 0, 0);
                    IncreaseDropRate(1);
                    break;
            }
        }


        //#################################################
        // GETTERS & SETTERS
        //#################################################
        public double BuildingDropRate
        {
            get { return _buildingDropRate; }
            set { _buildingDropRate = value; }
        }

        public int BuildingLevel
        {
            get { return _buildingLevel; }
            set { _buildingLevel = value; }
        }

        //#################################################
        // MEMBER VARIABLES
        //#################################################
        protected double _buildingDropRate;
        protected int _buildingLevel;
    }
}

File: Etap 4 - Implementacja C++/MiddleAges/building.cpp
#include "building.h"

//#################################################
// CONSTRUCTORS
//#################################################
Building::Building()
{

}

Building::Building(int priceGold, int priceWood, int priceStone, double dropRate = 1.0, int level = 1)
{
    setPriceGold(priceGold);
    setPriceWood(priceWood);
    setPriceStone(priceStone);
    setDropRate(dropRate);
    setLevel(level);
}

//#################################################
// MEMBER FUNCTIONS
//#################################################

void Building::increaseDropRate(double value)
{
    buildingDropRate *= value;
}

void Building::upgrade(int gold, int wood, int stone, double dropRate)
{
    buildingLevel += 1;
    switch (buildingLevel)
    {
    case 1:
        increasePrice(gold, wood, stone, 0, 0);
        increaseDropRate(1);
        break;
    case 2:
        increasePrice(1.1 * gold, 1.1 * wood, 1.1 * stone, 0, 0);
        increaseDropRate(1.1 * dropRate);
        break;
    case 3:
        increasePrice(1.2 * gold, 1.2 * wood, 1.2 * stone, 0, 0);
        increaseDropRate(1.2 * dropRate);
        break;
    case 4:
        increasePrice(1.5 * gold, 1.5 * wood, 1.5 * stone, 0, 0);
        increaseDropRate(1.3);
        break;
    case 5:
        increasePrice(1.8 * gold, 1.8 * wood, 1.8 * stone, 0, 0);
        increaseDropRate(1.4);
        break;
    default:
        increasePrice(0, 0, 0, 0, 0);
        increaseDropRate(1);
    }
}

//#################################################
// SETTERS
//#################################################
void Building::setDropRate(double dropRate)
{
    buildingDropRate = dropRate;
}

void Building::setLevel(int level)
{
    buildingLevel = level;
}

//#################################################
// GETTERS
//#################################################
double Building::getDropRate()
{
    return buildingDropRate;
}

int Building::getLevel()
{
    return buildingLevel;
}
